File: day4/c/main.c
#include <errno.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../util.h"

static const char *INPUT_FILENAME = "input.txt";

unsigned long filesize(FILE *f) {
  if(fseek(f, 0L, SEEK_END) == -1) {
    log_fatal("Failed to seek file. %s", strerror(errno));
  }

  long offset = ftell(f);
  if (offset == -1) {
    log_fatal("Failed to get size of file. %s", strerror(errno));
  }

  if (fseek(f, 0L, SEEK_SET) == -1) {
    log_fatal("Failed to reset file seek. %s", strerror(errno));
  }

  return (unsigned long) offset;
}

long buffer_size(unsigned long filesize) {
  int sigBit = sizeof(unsigned long)*8 - __builtin_clzl(filesize);
  
  return (int) pow(2.0, sigBit);
}

int matrix_coord(int r, int c, int rowSize) {
  return c + r*rowSize; 
}

int is_spool(int r, int c, int rowSize, int nRows, char *matrix) {
  if (r < 0 || r > nRows - 1) {
    return 0;
  }

  if (c < 0 || c > rowSize - 1) {
    return 0;
  }

  return matrix[matrix_coord(r, c, rowSize)] == '@';
}

int has_four_neighbors(int r, int c, int rowSize, int nRows, char *matrix) {

  int nNeighbors = 0;
  nNeighbors += is_spool(r + 1, c + 1, rowSize, nRows, matrix);
  nNeighbors += is_spool(r + 1, c, rowSize, nRows, matrix);
  nNeighbors += is_spool(r + 1, c - 1, rowSize, nRows, matrix);
  nNeighbors += is_spool(r , c + 1, rowSize, nRows, matrix);
  nNeighbors += is_spool(r , c - 1, rowSize, nRows, matrix);
  nNeighbors += is_spool(r - 1, c + 1, rowSize, nRows, matrix);
  nNeighbors += is_spool(r - 1, c , rowSize, nRows, matrix);
  nNeighbors += is_spool(r - 1, c - 1, rowSize, nRows, matrix);

  return nNeighbors > 3;
}

int clear_rolls(int rowSize, int nRows, char *matrix) {
  char *removeIndices[rowSize * nRows];
  int sum = 0;
  for (int r = 0; r < nRows; r++) {
    for (int c = 0; c < rowSize; c++) {
      int coord = matrix_coord(r, c, rowSize);
      if (matrix[coord] == '@' && !has_four_neighbors(r, c, rowSize, nRows, matrix)) {
        removeIndices[sum++] = &matrix[coord];
      }
    }
  }

  for (int i = 0; i < sum; i++) {
    *removeIndices[i] = '.';
  }

  return sum;

}

void p1() {
  FILE *f = fopen(INPUT_FILENAME, "r"); 
  if (f == NULL) {
    log_fatal("Could not open %s. %s", INPUT_FILENAME, strerror(errno));
  }

  long fsize = filesize(f);
  long bsize = buffer_size(fsize);
  char matrix[bsize];

  fread(matrix, sizeof(char), bsize, f);
  if (ferror(f)) {
    log_fatal("Failed to read from file %s. %s", INPUT_FILENAME, strerror(errno));
  }
  fclose(f);

  int rowSize = (int) (strchr(matrix, '\n') - matrix + 1); // first '\n' - matrix is the _index_ of the newline char, so plus one to get count.
  int nRows = fsize / (sizeof(char) * rowSize);

  long sum = 0;
  for (int r = 0; r < nRows; r++) {
    for (int c = 0; c < rowSize; c++) {
      if (matrix[matrix_coord(r, c, rowSize)] == '@' && !has_four_neighbors(r, c, rowSize, nRows, matrix)) {
        sum++;
      }
    }
  }

  printf("Part one: %ld\n", sum);
}

void p2() {
  FILE *f = fopen(INPUT_FILENAME, "r"); 
  if (f == NULL) {
    log_fatal("Could not open %s. %s", INPUT_FILENAME, strerror(errno));
  }

  long fsize = filesize(f);
  long bsize = buffer_size(fsize);
  char matrix[bsize];

  fread(matrix, sizeof(char), bsize, f);
  if (ferror(f)) {
    log_fatal("Failed to read from file %s. %s", INPUT_FILENAME, strerror(errno));
  }
  fclose(f);

  int rowSize = (int) (strchr(matrix, '\n') - matrix + 1); // first '\n' - matrix is the _index_ of the newline char, so plus one to get count.
  int nRows = fsize / (sizeof(char) * rowSize);

  long sum = 0;
  long numCleared = 1;
  while (numCleared != 0) {
    numCleared = clear_rolls(rowSize, nRows, matrix);
    sum += numCleared;
  }

  printf("Part two: %ld\n", sum);
}

int main() {
  p1();
  p2();
  
  return 0;
}

File: day4/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const INPUT_FILE = "input.txt"

func readMatrixFile(filename string) [][]byte {
	var input [][]byte = [][]byte{}

	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var read []byte = make([]byte, len(scanner.Bytes()))
		copy(read, scanner.Bytes())

		input = append(input, read)
	}

	var result [][]byte = make([][]byte, len(input)+2)
	for i := range result {
		result[i] = make([]byte, len(input[0])+2)
		for j := range result[i] {
			result[i][j] = '.'
		}
	}

	for i, slice := range input {
		copy(result[i+1][1:], slice)
	}

	return result
}

func hasNNeighbors(row, col, n int, matrix [][]byte) bool {
	var numNeigbors int = 0

	var neighborIndices [][]int = [][]int{
		{row + 1, col + 1},
		{row + 1, col},
		{row + 1, col - 1},
		{row, col + 1},
		{row, col - 1},
		{row - 1, col + 1},
		{row - 1, col},
		{row - 1, col - 1},
	}

	for _, neighbors := range neighborIndices {
		if matrix[neighbors[0]][neighbors[1]] == '@' {
			numNeigbors++
		}
	}

	return numNeigbors >= n
}

func clearRolls(matrix [][]byte) int {
	var removeIndices [][2]int = [][2]int{}

	var sum int = 0
	for row := 1; row < len(matrix)-1; row++ {
		for col := 1; col < len(matrix)-1; col++ {
			if matrix[row][col] == '@' && !hasNNeighbors(row, col, 4, matrix) {
				removeIndices = append(removeIndices, [2]int{row, col})
				sum += 1
			}
		}
	}

	var copyMatrix [][]byte = make([][]byte, len(matrix))
	for row := range matrix {
		copyMatrix[row] = make([]byte, len(matrix[row]))
		copy(copyMatrix[row], matrix[row])
	}

	for _, k := range removeIndices {
		matrix[k[0]][k[1]] = '.'
		copyMatrix[k[0]][k[1]] = 'x'
	}

	// printMatrix(copyMatrix)

	return sum
}

func printMatrix(matrix [][]byte) {
	for row := range matrix {
		for col := range matrix[row] {
			fmt.Printf("%s", string(matrix[row][col]))
		}
		fmt.Println()
	}
}

func min(x, y int) int {
	if x < y {
		return x
	}

	return x
}

func p1(filename string) {
	matrix := readMatrixFile(filename)

	copyMatrix := make([][]byte, len(matrix))
	for i := range matrix {
		copyMatrix[i] = make([]byte, len(matrix[i]))
		copy(copyMatrix[i], matrix[i])
	}

	var sum int = 0
	for row := 1; row < len(matrix)-1; row++ {
		for col := 1; col < len(matrix)-1; col++ {
			if matrix[row][col] == '@' && !hasNNeighbors(row, col, 4, matrix) {
				copyMatrix[row][col] = 'x'
				sum += 1
			}
		}
	}
	fmt.Printf("Part one: %d\n", sum)
}

func p2(filename string) {

	matrix := readMatrixFile(filename)
	// printMatrix(matrix)

	var sum int = 0
	var numCleared int = 1
	for numCleared != 0 {
		numCleared = clearRolls(matrix)

		sum += numCleared
	}

	fmt.Printf("Part two %d\n", sum)
}

func main() {
	p1(INPUT_FILE)
	p2(INPUT_FILE)
}
